include/syncmer.hpp

```diff
@@ -0,0 +1,567 @@
+    size_t w_size{};
+
+    //!\brief Stored values per window. It is necessary to store them, because a shift can remove the current syncmer.
+    std::deque<value_type> window_values{};
+
+    //!\brief Increments iterator by 1.
+    void next_unique_syncmer()
+    {
+        while (!next_syncmer()) {}
+    }
+
+    //!\brief Returns new window value.
+    auto window_value() const
+    {
+        return *urng1_iterator;
+    }
+
+
+    //!\brief Advances both windows to the next position.
+    void advance_window()
+    {
+        ++urng1_iterator;
+        ++urng2_iterator;
+    }
+
+    //!\brief Advances the first window to the next position.
+    void advance_first_window()
+    {
+        ++urng1_iterator;
+    }
+
+
+    //!\brief Calculates syncmers for the first window.
+    void window_first(const size_t kmer_size, const size_t smer_size)
+    {
+        w_size = kmer_size - smer_size + 1;
+
+        if (w_size == 0u)
+            return;
+
+        for (int i = 1u; i < kmer_size - 1 ; ++i)
+        {
+            window_values.push_back(window_value());
+            advance_first_window();
+        }
+        window_values.push_back(window_value());
+
+
+        auto smallest_s_it = std::ranges::min_element(window_values, std::less<value_type>{});
+        syncmer_position_offset = std::distance(std::begin(window_values), smallest_s_it);
+
+
+        if (opensyncmer && syncmer_position_offset == 0)
+        {
+            auto syncmer_it = urng2_iterator;
+            syncmer_value = *syncmer_it;
+        }
+
+        else if (!opensyncmer && syncmer_position_offset == 0 || syncmer_position_offset == w_size - 1 ) 
+        {
+            auto syncmer_it = urng2_iterator;
+            syncmer_value = *syncmer_it;
+        }
+
+    }
+
+    /*!\brief Calculates the next syncmer value.
+     * \returns True, if new syncmer is found or end is reached. Otherwise returns false.
+     * \details
+     * For the following windows, we remove the first window value (is now not in window_values) and add the new
+     * value that results from the window shifting.
+     */
+    bool next_syncmer()
+    {
+        advance_window();
+
+
+        if (urng1_iterator == urng1_sentinel)
+            return true;
+
+        value_type const new_value = window_value();
+
+        window_values.pop_front();
+        window_values.push_back(new_value);
+
+        if (syncmer_position_offset == 0)
+        {
+            auto smallest_s_it = std::ranges::min_element(window_values, std::less<value_type>{});
+            syncmer_position_offset = std::distance(std::begin(window_values), smallest_s_it);
+
+            if (opensyncmer && syncmer_position_offset == 0)
+            {
+                auto syncmer_it = urng2_iterator;
+                syncmer_value = *syncmer_it;
+                return true;
+            }
+
+            else if (!opensyncmer && syncmer_position_offset == 0 || syncmer_position_offset == w_size - 1 ) 
```

Comment:
Here too.

The idea is, if you use `if constexpr ()` then it is already processed during compile time and not during run time.